### PR TITLE
[Feature] Separate Creation of Example Submissions from Exercise Edit

### PR DESCRIPTION
**Pull request by Olimbia23**
### Checklist
- [X] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [X] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [X] Client: I added unit tests (Jest) related to the features (with a high test coverage)
- [X] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [X] Client: I documented the TypeScript code using JSDoc style.
- [X] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
Previously creation of example submissions was not intuitive for the instructors, because they had to go to edit exercise screen in order to add a new example submission.
Now I have provided a dedicated dashboard where the user can also create the example submissions.

### Description
Removed the creation of the example submissions from the exercise edit screen.
Provide an exercise example submissions table/dashboard for modelling and text exercises for both regular and exam.

### Steps for Testing (normal)
1. Log in to Artemis
2. Navigate to Course Management
3. Pick a course
4. Click on exercises
5. Create a text/modelling exercise
6. Click on example submissions

### Steps for Testing (exam)
1. Log in to Artemis
2. Navigate to Course Management
3. Pick a course
4. Click on exams
5. Create an exam
6. Create an exercise group
7. Create a text/modelling exercise
8. Click on example submissions

### Screenshots
Screenshot 2021-05-24 at 2 43 17 AM
Screenshot 2021-05-24 at 2 43 31 AM
Screenshot 2021-05-24 at 2 45 44 AM

### CODE REVIEW

**Review by joschahenningsen — CHANGES_REQUESTED** [review]
Generally works very well, nice work! 

When I update the explanation of the example submission (below the modeling editor), the request `PUT /api/modelling-submissions/123/example-assesment` fails with a 400 status:
```json
{"type":"https://www.jhipster.tech/problem/problem-with-message","title":"Bad Request","status":400,"detail":"Required request body is missing: public org.springframework.http.ResponseEntity<de.tum.in.www1.artemis.domain.Result> de.tum.in.www1.artemis.web.rest.ModelingAssessmentResource.saveModelingExampleAssessment(long,java.util.List<de.tum.in.www1.artemis.domain.Feedback>)","path":"/api/modeling-submissions/123/example-assessment","message":"error.http.400"}
```
This doesn't seem related to your changes, right?

When editing the modeling exercise there is only a "Create Example Solution" button even though there already is a example solution. It would be nice to see it or at least have some sort of indicator that there is one.

**Comment by Olimbia23** [comment]
@joschahenningsen Thank you for the review 😃 
> This doesn't seem related to your changes, right?

 Yes this PR has nothing to do with the update of example submission.

> When editing the modeling exercise there is only a "Create Example Solution" button even though there already is an example solution. It would be nice to see it or at least have some sort of indicator that there is one.

I think you are confusing _Example Solution_ with **Example Submissions**. 
There are many **Example Submissions** that an Instructor can create and this PR makes possible to create them through a dedicated example submissions dashboard. (previously the user had to create them on the Edit screen of an exercise)

However, this PR does not concern the creation of an _Example Solution._
There is only 1 _Example Solution_ an Instructor can create.
For Text Exercises this is done during the creation of the exercise itself. 
While for Modeling Exercise the process is more complicated since it is only possible to create it during the Edit of the Exercise. 
If @krusche  wants to separate this process like we did for example submissions, then we can discuss about it and create a new PR. 
After creating the example solution, the button is replaced by the link that redirects to the created example solution.(screenshots below)

Before creating the _example solution_: [edit screen for model exercise]
before

--------------------------------------------------------------------------------------------------

After creating the _example solution_: [edit screen for model exercise]
after

**Review by joschahenningsen — APPROVED** [review]
Thanks for clarifying @Olimbia23 I actually was a little confused by the terminology, In that case everything that's supposed to be covered by this works perfectly fine :ok_hand:

**Review by ArtemTsurikov — CHANGES_REQUESTED** [review]
Looks good, but the "Delete" button does not work and the error is shown on the "Exercises" page, not on the specific exercise example submission page.

Screenshot 2021-05-25 at 10 48 48

**Comment by Olimbia23** [comment]
> Looks good, but the "Delete" button does not work and the error is shown on the "Exercises" page, not on the specific exercise example submission page.

Hey @ArtemTsurikov thanks for the review. I have tested the delete button many times and it always worked fine for me. However, you are right about the alert being shown not in the right screen. I fixed that now.

**Review by willergeorg — DISMISSED** [review]
Tested on TS2. Didn't face the problems Artem had. Seems good for me.

**Review by NicolasUllmann — DISMISSED** [review]
Tested on Server 1. I also didn't run into the problem mentioned above, everything worked for me.

File: src/main/webapp/i18n/de/exercise.json [hunk 1]
```diff
@@ -189,7 +189,8 @@
                     "results": "Die bestehenden Ergebnisse müssen neu berechnet werden!",
                     "confirmtext": "Möchtest du die Änderungen wirklich übernehmen?"
                 }
-            }
+            },
+            "isUsedForTutorial": "Soll es vom Tutor benotet werden"
```
Comment (lschlesinger) [inline src/main/webapp/i18n/de/exercise.json:193]: I would translate it this way: Tutoren sollen diese Beispiellösung bewerten

File: src/main/webapp/i18n/de/global.json [hunk 2]
```diff
@@ -188,6 +188,7 @@
             "editInEditor": "In Editor bearbeiten",
             "editInIntelliJ": "In IntelliJ bearbeiten",
             "editEntity": "{{entity}} bearbeiten",
+            "exampleSubmissions": "Beispielabgaben",
```
Comment (lschlesinger) [inline src/main/webapp/i18n/de/global.json:191]: Is it Beispielabgaben or Beispiellösungen as in `src/main/webapp/i18n/de/exampleSubmission.json`?

File: src/main/webapp/i18n/en/exercise.json [hunk 3]
```diff
@@ -189,7 +189,8 @@
                     "results": "The existing results must be recalculated!",
                     "confirmtext": "Do you really want to perform these changes?"
                 }
-            }
+            },
+            "isUsedForTutorial": "Tutor needs to review"
```
Comment (lschlesinger) [inline src/main/webapp/i18n/en/exercise.json:193]: In the current text exercise I test, the checkbox says: "Tutors need to assess this submission".  I would change it to what is shown in the picture, it's the same use case?
image

**Review by lschlesinger — CHANGES_REQUESTED** [review]
I have some minor comments, the code looks good to me and I can approve the functionality!

**Review by ivanchimeno — APPROVED** [review]
Code looks good so I'm approving :) 

The only "issue" is: Should the example submissions really be accessible from the course management screen? Why not in the exercise detail page? Otherwise it would bloat up the tables with another button.

**Review by maximilianbergmann — DISMISSED** [review]
Tested on TS3. Works as expected

**Comment by Olimbia23** [comment]
> Code looks good so I'm approving :)
> 
> The only "issue" is: Should the example submissions really be accessible from the course management screen? Why not in the exercise detail page? Otherwise it would bloat up the tables with another button.

@ivanchimeno thank you for the review.
The button is also added to the exercise detail page. Screenshot is below
image

File: src/main/webapp/i18n/de/global.json (hunk 2 above)
Comment (Olimbia23) [reply src/main/webapp/i18n/de/global.json:191]: Thank you @lschlesinger for having a sharp eye on these terms.
I used Beispielabgaben for the example submissions.
Beispiellösung would be an **example solution**, which is another concept in Artemis, out of this PR scope.

File: src/main/webapp/i18n/en/exercise.json (hunk 3 above)
Comment (Olimbia23) [reply src/main/webapp/i18n/en/exercise.json:193]: Yes, you are correct, it is indeed the same use-case.
The dashboard column title is now the same key/string as shown on the screenshot.
But I renamed it a bit, by replacing the word _assess_ with _review_. This is the terminology Professor Brügge wants to be used by tutors from now on.

File: src/main/webapp/i18n/de/exercise.json (hunk 1 above)
Comment (Olimbia23) [reply src/main/webapp/i18n/de/exercise.json]: done

**Review by canyelten — DISMISSED** [review]
Tested on TS1, didn't face any problems.

**Review by ndricimrr — COMMENTED** [review]
Tested on TS1. Works as expected 👍

**Review by ndricimrr — DISMISSED** [review]

**Review by ivanchimeno — APPROVED** [review]

**Review by lschlesinger — DISMISSED** [review]
Tested on TS3. Worked fine!

**Review by NicolasUllmann — APPROVED** [review]
Tested on Server 1. Everything worked as expected.

**Review by willergeorg — DISMISSED** [review]
Still works.

File: src/main/webapp/app/exercises/modeling/manage/modeling-exercise-update.component.html [hunk 4]
```diff
@@ -222,51 +222,8 @@ <h4 *ngIf="isImport" id="jhi-modeling-exercise-heading-import" jhiTranslate="art
...
-            <button
-                *ngIf="modelingExercise.id && !isImport && isExamMode"
-                class="btn btn-primary"
-                [routerLink]="['/course-management', modelingExercise.exerciseGroup!.exam!.course!.id, 'modeling-exercises', modelingExercise.id, 'example-submissions', 'new']"
-                jhiTranslate="artemisApp.modelingExercise.createExampleSubmission"
-            >
-                Create example submission
-            </button>
-
             <div class="alert alert-info mt-3" *ngIf="!modelingExercise.id || isImport" jhiTranslate="artemisApp.modelingExercise.exampleSubmissionsRequireExercise">
```
Comment (FrankeLukas) [inline src/main/webapp/app/exercises/modeling/manage/modeling-exercise-update.component.html:227]: Unimportant nit: The `ngIf` here is no longer needed.

File: src/main/webapp/app/exercises/shared/exercise-detail-common-actions/non-programming-exercise-detail-common-actions.component.html [hunk 5]
```diff
@@ -24,6 +24,16 @@
         <span class="d-none d-md-inline" jhiTranslate="artemisApp.courseOverview.exerciseDetails.instructorActions.submissions">Submissions</span>
     </a>
 
+    <!-- Example Submission -->
+    <a
+        *ngIf="exercise.isAtLeastInstructor && (exercise.type == 'modeling' || exercise.type == 'text')"
```
Comment (FrankeLukas) [inline src/main/webapp/app/exercises/shared/exercise-detail-common-actions/non-programming-exercise-detail-common-actions.component.html:29]: Shouldn't this be allowed for editors too?

File: src/main/webapp/app/exercises/text/manage/text-exercise/text-exercise-row-buttons.component.html [hunk 6]
```diff
@@ -22,6 +22,15 @@
         <fa-icon [icon]="'book'"></fa-icon>
         <span class="d-none d-md-inline" jhiTranslate="artemisApp.courseOverview.exerciseDetails.instructorActions.submissions">Submissions</span>
     </a>
+    <!-- Example Submission -->
+    <a
+        *ngIf="exercise.isAtLeastInstructor"
```
Comment (FrankeLukas) [inline src/main/webapp/app/exercises/text/manage/text-exercise/text-exercise-row-buttons.component.html:27]: Same here.

File: src/main/webapp/app/exercises/shared/exam-exercise-row-buttons/exam-exercise-row-buttons.component.html [hunk 7]
```diff
@@ -35,6 +35,16 @@
         <span class="d-none d-md-inline" jhiTranslate="artemisApp.courseOverview.exerciseDetails.instructorActions.submissions">Submissions</span>
     </a>
 
+    <!-- Example Submission -->
+    <a
+        *ngIf="course.isAtLeastInstructor"
```
Comment (FrankeLukas) [inline src/main/webapp/app/exercises/shared/exam-exercise-row-buttons/exam-exercise-row-buttons.component.html:40]: I assume this should use the editor rule too. (As it looks the routes allow editors access when you enter the link directly anyway.)

File: src/main/webapp/app/exercises/modeling/manage/modeling-exercise.component.html [hunk 8]
```diff
@@ -89,6 +89,15 @@ <h4>
                                 <fa-icon [icon]="'book'"></fa-icon>
                                 <span class="d-none d-md-inline" jhiTranslate="artemisApp.courseOverview.exerciseDetails.instructorActions.submissions">Submissions</span>
                             </a>
+                            <!-- Example Submission -->
+                            <a
+                                *ngIf="modelingExercise.isAtLeastInstructor"
```
Comment (FrankeLukas) [inline src/main/webapp/app/exercises/modeling/manage/modeling-exercise.component.html:94]: Same here.

**Review by FrankeLukas — APPROVED** [review]
Can we remove the list or at least the links in the edit page? There are redundant to the new page.
grafik

Lgtm otherwise.

**Comment by FrankeLukas** [comment]
Ah yes, my bad. 🤦 Ignore that comment.

**Comment by jpbernius** [comment]
What Obstkernes status regarding the editor role comments from @FrankeLukas ?

File: src/main/webapp/app/exercises/shared/exam-exercise-row-buttons/exam-exercise-row-buttons.component.html (hunk 7 above)
Comment (Olimbia23) [reply src/main/webapp/app/exercises/shared/exam-exercise-row-buttons/exam-exercise-row-buttons.component.html]: done.

File: src/main/webapp/app/exercises/shared/exercise-detail-common-actions/non-programming-exercise-detail-common-actions.component.html (hunk 5 above)
Comment (Olimbia23) [reply src/main/webapp/app/exercises/shared/exercise-detail-common-actions/non-programming-exercise-detail-common-actions.component.html]: yes, done.

**Comment by Olimbia23** [comment]
> What is the status regarding the editor role comments from @FrankeLukas ?

I incorporated his feedback. @jpbernius

**Review by jpbernius — APPROVED** [review]

**Pull request merged**